### PR TITLE
feat: add erc20 and erc721 token wrappers

### PR DESCRIPTION
**Pull request by adrianhdez929**

### CODE REVIEW

**Comment by adrianhdez929** [comment]
@TheSnakeWitcher could you please review this to merge and release? Its needed already on another repo :)

File: test/Wrappers/wrapper-erc20.test.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+        expect(await tokenWrapper.symbol())
+            .to.be.equal("MKT")
+        expect(await tokenWrapper.name())
+            .to.be.equal("Mock")
+        expect(await tokenWrapper.decimals())
+            .to.be.equal("18")
+        expect(await tokenWrapper.totalSupply())
+            .to.be.equal("100")
+    })
```
Comment (TheSnakeWitcher) [inline test/Wrappers/wrapper-erc20.test.ts:28]: To make every `it` block and actual unit test try to test a single property in each one

File: test/Wrappers/wrapper-erc20.test.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    it("should transfer", async () => {
+        const [deployer, target] = await ethers.getSigners()
+
+        const tokenWrapper = new ERC20Wrapper(
+            tokenContract.target, ethers.provider
+        )
```
Comment (TheSnakeWitcher) [inline test/Wrappers/wrapper-erc20.test.ts:34]: If wrapper it is needed try to create it in a `beforeEach()` hook and ideally with a fixture, [here an example](https://github.com/Fusyona/web3-core/blob/7de6dc7ca04f0db322b13211913db41f21761940/test/Cooldown.ts?plain=1#L8-L21)

File: test/Wrappers/wrapper-erc20.test.ts [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        expect(await tokenWrapper.balanceOf(await target.getAddress()))
+            .to.be.equal("0")
+
+        await tokenWrapper.withSigner(deployer).transfer(await target.getAddress(), "10")
+
+        expect(await tokenWrapper.balanceOf(await deployer.getAddress()))
+            .to.be.equal("90")
+        expect(await tokenWrapper.balanceOf(await target.getAddress()))
+            .to.be.equal("10")
+    })
```
Comment (TheSnakeWitcher) [inline test/Wrappers/wrapper-erc20.test.ts:47]: Here we could use [changetokenbalance](https://hardhat.org/hardhat-chai-matchers/docs/reference#.changetokenbalance)

File: test/Wrappers/wrapper-erc20.test.ts [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        expect(await tokenWrapper.allowance(await deployer.getAddress(), await spender.getAddress()))
+            .to.be.equal("10")
+
+        await tokenWrapper.withSigner(spender).transferFrom(await deployer.getAddress(), await target.getAddress(), "10")
+
+        expect(await tokenWrapper.balanceOf(await deployer.getAddress()))
+            .to.be.equal("90")
+        expect(await tokenWrapper.balanceOf(await target.getAddress()))
+            .to.be.equal("10")
+    })
```
Comment (TheSnakeWitcher) [inline test/Wrappers/wrapper-erc20.test.ts:71]: Here we could use [changetokenbalance](https://hardhat.org/hardhat-chai-matchers/docs/reference#.changetokenbalance)

File: test/Wrappers/wrapper-erc20.test.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import { ERC20Mock } from "../../typechain-types"
+
+describe("ERC20 Wrapper", async () => {
+    let tokenContract: ERC20Mock
+
+    beforeEach(async () => {
+        const [deployer] = await ethers.getSigners()
+
+        tokenContract = await ethers.deployContract("ERC20Mock", [deployer, 100, "Mock", "MKT"])
+    })
```
Comment (TheSnakeWitcher) [inline test/Wrappers/wrapper-erc20.test.ts:13]: Directly create the wrapper here, if we need the contract you can use `wrapper.call` and if you need the contract address we can use `wrapper.address`

**Review by TheSnakeWitcher — CHANGES_REQUESTED** [review]
@adrianhdez929 Nice work, I just point a few minor details, mainly to avoid the change in the structure, at least for now to avoid breaking things,note that all suggestions of `erc20-wrapper` applies too for `erc721-wrapper`, keep nice work

File: test/Wrappers/wrapper-erc20.test.ts (hunk 3 above)
Comment (adrianhdez929) [reply test/Wrappers/wrapper-erc20.test.ts:47]: The idea there is to use the balanceOf function from the wrapper

Comment (TheSnakeWitcher) [reply test/Wrappers/wrapper-erc20.test.ts:50]: There is no problem in these way, but I found these approach a little more verbose. The `balanceOf()` function should be tested in its own `it` block for it

Comment (adrianhdez929) [reply test/Wrappers/wrapper-erc20.test.ts]: Done! It was more intuitive to check `balanceOf` just at the start of the test and use `changeTokenBalances` after a transfer

File: lib/main.ts [hunk 6]
```diff
@@ -1,3 +1,3 @@
 export * from "./types";
 export * from "./components";
-export { default as BaseWrapper } from "./base-wrapper";
+export { default as BaseWrapper } from "./wrappers/base-wrapper";
```
[the remark this answers is not in the document]
Comment (adrianhdez929) [reply lib/main.ts:3]: Roger that, I will keep the base-wrapper file on the original path, but will keep the folder for incoming wrapper implementations

[the remark this answers is not in the document]
Comment (TheSnakeWitcher) [reply lib/main.ts]: Nice 👍🏿

File: lib/wrappers/tokens/erc20.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    ) {
+        const contract = new Contract(address, ERC20ABI, provider)
+        super(contract, provider)
+    }
+
+    async totalSupply(): Promise<bigint> {
+        return this.call.totalSupply()
+    }
+
+    async balanceOf(account: Address): Promise<bigint> {
```
Comment (MakeMake23) [inline lib/wrappers/tokens/erc20.ts:27]: Why all these methods are necessary if they just use `this.call` directly?

File: lib/wrappers/tokens/erc20.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        return this.call.balanceOf(account)
+    }
+
+    async transfer(to: Address, value: string) {
```
Comment (MakeMake23) [inline lib/wrappers/tokens/erc20.ts:31]: You should wait a number of confirmations before returning. I think there is a helper for that in the base class.

File: lib/wrappers/tokens/erc721.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
+import {
```
Comment (MakeMake23) [inline lib/wrappers/tokens/erc721.ts:1]: Same comments above apply.

File: test/Wrappers/wrapper-erc721.test.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+        expect(await tokenWrapper.balanceOf(await deployer.getAddress()))
+            .to.be.equal("1")
+        expect(await tokenWrapper.balanceOf(await target.getAddress()))
+            .to.be.equal("0")
+
+        expect(await tokenWrapper.withSigner(deployer).transferFrom(await deployer.getAddress(), await target.getAddress(), "1"))
+            .to.changeTokenBalance(tokenContract, [deployer, target], [-1, 1])
+    })
+    it("should approve and approveAll and transferFrom", async () => {
```
Comment (MakeMake23) [inline test/Wrappers/wrapper-erc721.test.ts:43]: Try keeping tests as atomic as possible. An ideal test has only one `expect` statement.

**Review by MakeMake23 — CHANGES_REQUESTED** [review]
Good job @adrianhdez929. Hope my comments are helpful.

File: lib/wrappers/tokens/erc20.ts (hunk 7 above)
Comment (adrianhdez929) [reply lib/wrappers/tokens/erc20.ts:27]: Are you suggesting all the non-mutable methods should go inside the "call" abstraction? And just expose those that actually transact?

File: lib/main.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheSnakeWitcher) [reply lib/main.ts]: Fixed in 29f39db

File: test/Wrappers/wrapper-erc20.test.ts (hunk 1 above)
Comment (TheSnakeWitcher) [reply test/Wrappers/wrapper-erc20.test.ts]: Fixedi in 118f2aa

File: test/Wrappers/wrapper-erc20.test.ts (hunk 3 above)
Comment (TheSnakeWitcher) [reply test/Wrappers/wrapper-erc20.test.ts]: Fixed in d29f2d1

File: test/Wrappers/wrapper-erc20.test.ts (hunk 4 above)
Comment (TheSnakeWitcher) [reply test/Wrappers/wrapper-erc20.test.ts]: Fixed in d29f2d1

File: test/Wrappers/wrapper-erc20.test.ts (hunk 5 above)
Comment (TheSnakeWitcher) [reply test/Wrappers/wrapper-erc20.test.ts]: Fixed in 118f2aa

File: lib/wrappers/tokens/erc20.ts (hunk 8 above)
Comment (TheSnakeWitcher) [reply lib/wrappers/tokens/erc20.ts:24]: Fixed in ab9fd61

File: lib/wrappers/tokens/erc20.ts (hunk 7 above)
Comment (TheSnakeWitcher) [reply lib/wrappers/tokens/erc20.ts]: Fixed in 3e64ae7

File: test/Wrappers/wrapper-erc721.test.ts (hunk 10 above)
Comment (TheSnakeWitcher) [reply test/Wrappers/wrapper-erc721.test.ts]: Fixed in b6c571d

File: lib/wrappers/tokens/erc721.ts (hunk 9 above)
Comment (TheSnakeWitcher) [reply lib/wrappers/tokens/erc721.ts:1]: Fixed in ab9fd61

**Pull request merged**